Add support to use URI mapper with custom metrics recorder

fkjellberg opened the pull request:
Add support to use URI mapper function together with custom metrics recorder. It's more efficient to let HttpClientMetricsHandler and HttpServerMetricsHandler call the mapper function before calling the recorder methods compared to do the mapping in the custom metrics recorder since the handlers will reuse an already mapped value when calling different methods in the recorder.

pivotal-issuemaster commented:
@fkjellberg Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/reactor?repositoryId=reactor/reactor-netty&pullRequestId=1297)!

[Click here](https://cla.pivotal.io/sync/reactor?repositoryId=reactor/reactor-netty&pullRequestId=1297) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-issuemaster commented:
@fkjellberg Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/reactor?repositoryId=reactor/reactor-netty&pullRequestId=1297)!

violetagg commented on reactor-netty-http/src/main/java/reactor/netty/http/server/HttpServer.java:
```diff
@@ -315,14 +315,23 @@ public final HttpServer httpRequestDecoder(Function<HttpRequestDecoderSpec, Http
+
+	public final HttpServer metrics(boolean enable, Supplier<? extends ChannelMetricsRecorder> recorder, Function<String, String> uriTagValue) {
 		if (enable) {
 			if (!Metrics.isInstrumentationAvailable()) {
 				throw new UnsupportedOperationException(
 						"To enable metrics, you must add the dependency `io.micrometer:micrometer-core`" +
 								" to the class path first");
 			}
```
Do we need to check for Micrometer as this is a custom recorder that might not need it?
We need it only in case of the default implementation.

fkjellberg replied:
No, you are right and I now understand why the code looked like it did before my change. I'll fix.

violetagg commented on reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClient.java:
```diff
@@ -1105,14 +1105,23 @@ public final HttpClient mapConnect(
+
+	public final HttpClient metrics(boolean enable, Supplier<? extends ChannelMetricsRecorder> recorder, Function<String, String> uriTagValue) {
 		if (enable) {
 			if (!Metrics.isInstrumentationAvailable()) {
 				throw new UnsupportedOperationException(
 						"To enable metrics, you must add the dependency `io.micrometer:micrometer-core`" +
 								" to the class path first");
 			}
```
Do we need to check for Micrometer as this is a custom recorder that might not need it?
We need it only in case of the default implementation.

fkjellberg replied:
I'll fix here as well

violetagg reviewed (changes requested):
Can you add a test for the new functionality?

Do you need this feature only for 1.0.0 (still in development) or 0.9.x (the current version with releases)?

fkjellberg commented:
@violetagg I fixed your requested changes and added two tests, one for HttpClient and one for HttpServer.

I need this functionality in the Spring Cloud Gateway so it would be great if this can be backported to the 0.9.x branch. Do you want me to create a separate PR for that?

violetagg commented:
> @violetagg I fixed your requested changes and added two tests, one for HttpClient and one for HttpServer.
> 
> I need this functionality in the Spring Cloud Gateway so it would be great if this can be backported to the 0.9.x branch. Do you want me to create a separate PR for that?

OK I'll mark this PR for 1.0.0.RC1. Once we agree on the change, if you are able to create a PR for 0.9.x it will be great.

fkjellberg commented:
I noticed that the test cases failed in some of the automatic builds. After struggling a bit they now pass after my latest push.

violetagg commented on reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClient.java:
```diff
@@ -1132,6 +1132,24 @@ public final HttpClient metrics(boolean enable, Supplier<? extends ChannelMetric
 		return super.metrics(enable, recorder);
 	}
 
+	public final HttpClient metrics(boolean enable, Supplier<? extends ChannelMetricsRecorder> recorder, Function<String, String> uriTagValue) {
```
Add javadoc

violetagg commented on reactor-netty-http/src/main/java/reactor/netty/http/server/HttpServer.java:
```diff
@@ -342,6 +342,24 @@ public final HttpServer metrics(boolean enable, Supplier<? extends ChannelMetric
 		return super.metrics(enable, recorder);
 	}
 
+	public final HttpServer metrics(boolean enable, Supplier<? extends ChannelMetricsRecorder> recorder, Function<String, String> uriTagValue) {
```
Add javadoc

fkjellberg commented:
@violetagg PR updated with javadoc but could you please give me some advice regarding the unit tests as one of them again fails on one of the build bots? I took inspiration from other similar test cases but apparently I'm missing something?

violetagg commented on reactor-netty-http/src/test/java/reactor/netty/http/client/HttpClientTest.java:
```diff
@@ -2489,4 +2491,88 @@ public void testLoopAndResolver() {
+							.wiretap(true)
+							.bindNow();
+
+		final List<String> collectedUris = Collections.synchronizedList(new ArrayList<>());
```
Instead of this can we use `Sinks`
`Sinks.Many<String> collectedUris = Sinks.many().replay().all();`

violetagg commented on reactor-netty-http/src/test/java/reactor/netty/http/client/HttpClientTest.java:
```diff
@@ -2489,4 +2491,88 @@ public void testLoopAndResolver() {
+						() -> new HttpClientMetricsRecorder() {
+							@Override
+							public void recordDataReceived(SocketAddress remoteAddress, String uri, long bytes) {
+								collectedUris.add(uri);
```
`collectedUris.emitNext(uri);`

violetagg commented on reactor-netty-http/src/test/java/reactor/netty/http/client/HttpClientTest.java:
```diff
@@ -2489,4 +2491,88 @@ public void testLoopAndResolver() {
+
+							@Override
+							public void incrementErrorsCount(SocketAddress remoteAddress, String uri) {
+								collectedUris.add(uri);
```
`collectedUris.emitNext(uri);`

violetagg commented on reactor-netty-http/src/test/java/reactor/netty/http/client/HttpClientTest.java:
```diff
@@ -2489,4 +2491,88 @@ public void testLoopAndResolver() {
+
+							@Override
+							public void recordDataSent(SocketAddress remoteAddress, String uri, long bytes) {
+								collectedUris.add(uri);
```
`collectedUris.emitNext(uri);`

violetagg commented on reactor-netty-http/src/test/java/reactor/netty/http/client/HttpClientTest.java:
```diff
@@ -2489,4 +2491,88 @@ public void testLoopAndResolver() {
+
+							@Override
+							public void recordDataReceivedTime(SocketAddress remoteAddress, String uri, String method, String status, Duration time) {
+								collectedUris.add(uri);
```
`collectedUris.emitNext(uri);`

violetagg commented on reactor-netty-http/src/test/java/reactor/netty/http/client/HttpClientTest.java:
```diff
@@ -2489,4 +2491,88 @@ public void testLoopAndResolver() {
+
+							@Override
+							public void recordDataSentTime(SocketAddress remoteAddress, String uri, String method, Duration time) {
+								collectedUris.add(uri);
```
`collectedUris.emitNext(uri);`

violetagg commented on reactor-netty-http/src/test/java/reactor/netty/http/client/HttpClientTest.java:
```diff
@@ -2489,4 +2491,88 @@ public void testLoopAndResolver() {
+
+							@Override
+							public void recordResponseTime(SocketAddress remoteAddress, String uri, String method, String status, Duration time) {
+								collectedUris.add(uri);
```
`collectedUris.emitNext(uri);`

violetagg commented on reactor-netty-http/src/test/java/reactor/netty/http/client/HttpClientTest.java:
```diff
@@ -2489,4 +2491,88 @@ public void testLoopAndResolver() {
+				.aggregate()
+				.block(Duration.ofSeconds(30));
+
+		synchronized (collectedUris) {
+			assertThat(collectedUris).isNotEmpty();
+			for(String uri : collectedUris) {
+				assertThat(uri).isEqualTo("/stream/{n}");
+			}
+		}
```
```
		List<String> expected = IntStream.range(0, 5)
				.mapToObj(i -> "/stream/{n}")
				.collect(Collectors.toList());
		StepVerifier.create(collectedUris.asFlux().take(5))
				.expectNextSequence(expected)
				.expectComplete()
				.verify();
```

violetagg reviewed (changes requested):
Can we use `Sinks` for the test?

violetagg commented:
> @violetagg PR updated with javadoc but could you please give me some advice regarding the unit tests as one of them again fails on one of the build bots? I took inspiration from other similar test cases but apparently I'm missing something?

We can use `Sinks`. As it is a new API in Reactor Core, when you create the PR for 0.9.x, the corresponding type is `ReplayProcessor`

fkjellberg commented:
@violetagg Thanks for the help! PR updated with your proposed changes.

simonbasle commented on reactor-netty-http/src/test/java/reactor/netty/http/client/HttpClientTest.java:
```diff
@@ -2489,4 +2493,89 @@ public void testLoopAndResolver() {
+							.wiretap(true)
+							.bindNow();
+
+		Sinks.Many<String> collectedUris = Sinks.many().replay().all();
```
I see little sense in using a Sink/Flux here, a `List` (make it a `CopyOnWriteArrayList` if you're concerned about concurrency) would be perfectly fine here.

violetagg replied:
@simonbasle The PR used List, but with List it is not guaranteed that exactly after receiving the response  all values will be in the least (see previous commits and the failed CI). So we either use `Sinks` or other mechanism (CountDownLatch) in order to be sure that when we start checking the values all data is there.

We can use CoundDownLatch if this is your recommendation.

simonbasle replied:
I see. I would have expected the metrics to be recorded by the time the client`.get()` is finished (since it is blocking). I found the pattern of using `Sinks` a bit surprising and harder to read than a list-based assertion in that particular case.

Using `take(5)` is a potential subtle issue, too. It wouldn't catch a bug where a 6th uri is recorded.

So my preference would be to switch back to a `List`+`assertJ` approach. reactor-netty tests don't depend on `Awaitility` from what I can see, but it could be a better alternative than `CountdownLatch` here, wdyt @violetagg ?

Plus the assertion would be simpler and be less brittle:

```java
assertThat(collectedUris).hasSize(5).containsOnly("/stream/{n}");
```

Either way is fine in the end, I'll leave that decision to you @violetagg. I would just add a clarifying comment to the use of `Sinks` if we keep this solution.

violetagg replied:
the number `5` is because we have 5 types of metrics and not 5 requests, because of that we cannot expect on `get` to have all of them.
For one request that we make we have `recordDataReceived`, `recordDataSent`, `recordDataReceivedTime`, `recordDataSentTime`, `recordResponseTime`

simonbasle replied:
yeah, I understood that. I meant, if a bug is introduced in the `MetricsRecorder` (or the internal code that calls it) and ends up calling eg. `recordDataReceived` twice, this test wouldn't see it (due to the `take(5)`) whereas a list assertion like I suggested would.

violetagg replied:
@fkjellberg Please switch to `CountDownLatch` or `Awaitility` instead of `Sinks`

simonbasle commented on reactor-netty-http/src/test/java/reactor/netty/http/client/HttpClientTest.java:
```diff
@@ -2489,4 +2493,89 @@ public void testLoopAndResolver() {
+							}
+
+							@Override
+							public void recordDataReceived(SocketAddress remoteAddress, long bytes) {
```
@violetagg do we have an abstract/no-op base class (either in prod or test sources) that could remove the need to implement all these empty methods?

violetagg replied:
no there isn't

simonbasle commented on reactor-netty-http/src/test/java/reactor/netty/http/server/HttpServerTests.java:
```diff
@@ -2109,4 +2111,88 @@ else if (msg instanceof ByteBuf) {
+
+	@Test
+	public void testCustomMetricsWithUriMapper() {
+		Sinks.Many<String> collectedUris = Sinks.many().replay().all();
```
same as above

simonbasle commented on reactor-netty-http/src/test/java/reactor/netty/http/server/HttpServerTests.java:
```diff
@@ -2109,4 +2111,88 @@ else if (msg instanceof ByteBuf) {
+										@Override
+										public void recordDataReceived(SocketAddress remoteAddress, String uri, long bytes) {
+											collectedUris.emitNext(uri);
+										}
```
same as above

simonbasle reviewed (changes requested):
nitpicks on the tests. also @fkjellberg I'm not sure I understand your comment about efficiency, since I don't see any change in the way the mapper is applied in this PR, can you clarify?

simonbasle reviewed (changes requested):
nitpicks on the tests. also @fkjellberg I'm not sure I understand your comment about efficiency, since I don't see any change in the way the mapper is applied in this PR, can you clarify?

violetagg commented:
> nitpicks on the tests. also @fkjellberg I'm not sure I understand your comment about efficiency, since I don't see any change in the way the mapper is applied in this PR, can you clarify?

When there is `uriTagValue`, Reactor Netty calculates the path once (when the request is received)
https://github.com/reactor/reactor-netty/blob/3528badfa032195e8325f35ba351e993cf186a9a/reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClientMetricsHandler.java#L70

without `uriTagValue` for the custom recorders, they (the recorders) has to do this on every invocation.

simonbasle commented:
> When there is `uriTagValue`, Reactor Netty calculates the path once (when the request is about to be sent)

Thanks for clarifying that for me. So in essence the "plumbing" was there but the API didn't guide you into setting the pattern along activation of the metrics. 👍

violetagg commented on reactor-netty-http/src/test/java/reactor/netty/http/server/HttpServerTests.java:
```diff
@@ -2109,4 +2111,87 @@ else if (msg instanceof ByteBuf) {
+							.bindNow();
+
+		HttpClient.create()
+				.doOnRequest((req, conn) -> conn.onTerminate()
```
It is better to count down when every metric is recorded, thus `latch.await` will wait for all to appear.
Wdyt?

simonbasle replied:
like initializing as `CountDownLatch(5)` and counting down next to each `collectedUris.add()` call?

why not, especially if you think counting down on the connection termination could lead to making the assertion too early. it has the same drawback as my original concern with `take(5)`, but that is more minor.

I like how the test reads much more with the list 👍

violetagg replied:
yes because terminating is competing with recorder
https://github.com/reactor/reactor-netty/blob/8afc89e173733e97ce67d8bcbf4e3083e8a40f1f/reactor-netty-http/src/main/java/reactor/netty/http/server/HttpTrafficHandler.java#L280
https://github.com/reactor/reactor-netty/blob/8afc89e173733e97ce67d8bcbf4e3083e8a40f1f/reactor-netty-http/src/main/java/reactor/netty/http/server/HttpServerMetricsHandler.java#L77

fkjellberg commented:
@simonbasle @violetagg Thanks for the feedback! Please review the updated PR where the test cases are now using CountDownLatch. One of the build bots failed but it seems unrelated to my changes as it was in the DefaultPooledConnectionProviderTest.

violetagg commented:
> @simonbasle @violetagg Thanks for the feedback! Please review the updated PR where the test cases are now using CountDownLatch. One of the build bots failed but it seems unrelated to my changes as it was in the DefaultPooledConnectionProviderTest.

I do not see the difference between the last two commits?

violetagg commented:
@fkjellberg I squashed the changes and committed them with 6a79577. Also I backported the change to 0.9.x with 96da5d0.
Thanks a lot for the PR.
0.9.12.RELEASE and 1.0.0-RC1 are scheduled for today.

fkjellberg commented:
Thank you, @violetagg ! Looking forward to get this into the next Spring Cloud Gateway release.
